File: DemoServer/AdditionalLanguages/go/Queries/sortingQueryResults/SortingQueryResults.go
package main

//region Usings
import "github.com/ravendb/ravendb-go-client"
//endregion

var globalDocumentStore *ravendb.DocumentStore

func main() {
    createDocumentStore()
    createDatabase()
    sortingQueryResults(20)
    globalDocumentStore.Close()
}

func createDocumentStore() (*ravendb.DocumentStore, error) {
    if globalDocumentStore != nil {
        return globalDocumentStore, nil
    }
    urls := []string{"http://localhost:8080"}
    store := ravendb.NewDocumentStore(urls, "testGO")
    err := store.Initialize()
    if err != nil {
        return nil, err
    }
    globalDocumentStore = store
    return globalDocumentStore, nil
}

func createDatabase() {
    databaseRecord := ravendb.NewDatabaseRecord()
    databaseRecord.DatabaseName = "testGO"
    createDatabaseOperation := ravendb.NewCreateDatabaseOperation(databaseRecord, 1)
    var err = globalDocumentStore.Maintenance().Server().Send(createDatabaseOperation)
    if err != nil {
        fmt.Printf("d.store.Maintenance().Server().Send(createDatabaseOperation) failed with %s\n", err)
    }
}

//region Demo
func sortingQueryResults(numberOfUnits int) error {

    session, err := globalDocumentStore.OpenSession("")
    if err != nil {
        return err
    }
    defer session.Close()

    //region Step_1
    queriedType := reflect.TypeOf(&Product{})
    query := session.QueryCollectionForType(queriedType)
    //endregion

    //region Step_2
    query = query.Where("UnitsInStock", ">", numberOfUnits)
    //endregion
    //region Step_3
    query = query.OrderByDescendingWithOrdering("UnitsInStock", ravendb.OrderingTypeLong)
    //endregion
    //region Step_4
    query = query.OrderByWithOrdering("Name", ravendb.OrderingTypeAlphaNumeric)
    //endregion

    //region Step_5
    var sortedProducts []*Product
    err = query.GetResults(&sortedProducts)
    if err != nil {
        return err
    }
    //endregion

    return nil
}

type Product struct {
    ID              string
    Name            string
    Supplier        string
    Category        string
    QuantityPerUnit string
    PricePerUnit    float64
    UnitsInStock    int 
    UnitsOnOrder    int
    Discontinued    bool 
    ReorderLevel    int 
}
//endregion

File: DemoServer/AdditionalLanguages/java/src/main/java/net/ravendb/demo/queries/sortingQueryResults/SortingQueryResults.java
package net.ravendb.demo.queries.sortingQueryResults;

//region Usings
import net.ravendb.client.documents.session.IDocumentSession;
import net.ravendb.client.documents.session.OrderingType;
import java.util.List;
//endregion
import net.ravendb.demo.common.DocumentStoreHolder;
import net.ravendb.demo.common.models.Product;

public class SortingQueryResults {
    public List<Product> run(RunParams runParams) {
        double numberOfUnits = runParams.getNumberOfUnits();

        //region Demo
        List<Product> sortedProducts;

        try (IDocumentSession session = DocumentStoreHolder.store.openSession()) {
            //region Step_1
            sortedProducts = session.query(Product.class)
            //endregion
                //region Step_2
                .whereGreaterThan("UnitsInStock", numberOfUnits)
                //endregion
                //region Step_3
                .orderByDescending("UnitsInStock")
                //endregion
                //region Step_4
                .orderBy("Name", OrderingType.ALPHA_NUMERIC)
                //endregion
                //region Step_5
                .toList();
                //endregion
        }
        //endregion

        return sortedProducts;
    }

    public static class RunParams {
        private double numberOfUnits;

        public double getNumberOfUnits() {
            return numberOfUnits;
        }

        public void setNumberOfUnits(double numberOfUnits) {
            this.numberOfUnits = numberOfUnits;
        }
    }
}
